File: packages/sdk-typescript/src/instrumentations/mcp.ts
/**
 * MCP (Model Context Protocol) instrumentation for Tracelane.
 *
 * Wraps MCP Client's callTool() and readResource() methods to emit OTel
 * spans. MCP is the most-attacked agent surface; these spans feed
 * Tracelane's MCP tool-description hash watcher (PP-PR1) and provide
 * the observability layer for rug-pull detection.
 *
 * @example
 * ```ts
 * import { Client } from "@modelcontextprotocol/sdk/client/index.js";
 * import { instrumentMCP } from "@tracelanedev/sdk/mcp";
 *
 * const client = new Client({ name: "my-client", version: "1.0.0" });
 * instrumentMCP(client);
 * const result = await client.callTool({ name: "my_tool", arguments: {} });
 * ```
 */

import { SpanKind, SpanStatusCode, trace } from "@opentelemetry/api";

const tracer = trace.getTracer("@tracelanedev/sdk-mcp", "0.1.0");

interface MCPClientLike {
	callTool: (...args: unknown[]) => Promise<unknown>;
	readResource?: (...args: unknown[]) => Promise<unknown>;
	listTools?: (...args: unknown[]) => Promise<unknown>;
}

/**
 * Instrument an MCP Client instance to emit OTel spans.
 *
 * @param client - An @modelcontextprotocol/sdk Client or ClientSession instance
 */
export function instrumentMCP(client: MCPClientLike): void {
	_patchCallTool(client);
	if (client.readResource) {
		_patchReadResource(
			client as MCPClientLike & {
				readResource: (...args: unknown[]) => Promise<unknown>;
			},
		);
	}
	if (client.listTools) {
		_patchListTools(
			client as MCPClientLike & {
				listTools: (...args: unknown[]) => Promise<unknown>;
			},
		);
	}
}

function _patchCallTool(client: MCPClientLike): void {
	const originalCallTool = client.callTool.bind(client);

	client.callTool = async (...args: unknown[]) => {
		const request = args[0] as Record<string, unknown> | undefined;
		const toolName =
			typeof request?.name === "string" ? request.name : "unknown";
		const argCount =
			request?.arguments && typeof request.arguments === "object"
				? Object.keys(request.arguments as object).length
				: 0;

		return tracer.startActiveSpan(
			"mcp.callTool",
			{
				kind: SpanKind.CLIENT,
				attributes: {
					"gen_ai.provider.name": "mcp",
					"mcp.tool_name": toolName,
					"mcp.argument_count": argCount,
				},
			},
			async (span) => {
				try {
					const result = (await originalCallTool(...args)) as Record<
						string,
						unknown
					>;
					const isError = Boolean(result.isError ?? false);
					span.setAttribute("mcp.is_error", isError);
					const content = result.content;
					if (Array.isArray(content)) {
						span.setAttribute("mcp.content_count", content.length);
					}
					span.setStatus({
						code: isError ? SpanStatusCode.ERROR : SpanStatusCode.OK,
						message: isError ? "MCP tool returned isError=true" : "",
					});
					return result;
				} catch (e) {
					span.recordException(e as Error);
					span.setStatus({ code: SpanStatusCode.ERROR, message: String(e) });
					throw e;
				} finally {
					span.end();
				}
			},
		);
	};
}

function _patchReadResource(
	client: MCPClientLike & {
		readResource: NonNullable<MCPClientLike["readResource"]>;
	},
): void {
	const originalReadResource = client.readResource.bind(client);

	client.readResource = async (...args: unknown[]) => {
		const request = args[0] as Record<string, unknown> | undefined;
		const uri = typeof request?.uri === "string" ? request.uri : "unknown";

		return tracer.startActiveSpan(
			"mcp.readResource",
			{
				kind: SpanKind.CLIENT,
				attributes: {
					"gen_ai.provider.name": "mcp",
					"mcp.resource_uri": uri,
				},
			},
			async (span) => {
				try {
					const result = (await originalReadResource(...args)) as Record<
						string,
						unknown
					>;
					const contents = result.contents;
					if (Array.isArray(contents)) {
						span.setAttribute("mcp.content_count", contents.length);
					}
					span.setStatus({ code: SpanStatusCode.OK });
					return result;
				} catch (e) {
					span.recordException(e as Error);
					span.setStatus({ code: SpanStatusCode.ERROR, message: String(e) });
					throw e;
				} finally {
					span.end();
				}
			},
		);
	};
}

function _patchListTools(
	client: MCPClientLike & {
		listTools: NonNullable<MCPClientLike["listTools"]>;
	},
): void {
	const originalListTools = client.listTools.bind(client);

	client.listTools = async (...args: unknown[]) => {
		return tracer.startActiveSpan(
			"mcp.listTools",
			{
				kind: SpanKind.CLIENT,
				attributes: { "gen_ai.provider.name": "mcp" },
			},
			async (span) => {
				try {
					const result = (await originalListTools(...args)) as Record<
						string,
						unknown
					>;
					const tools = result.tools;
					if (Array.isArray(tools)) {
						span.setAttribute("mcp.tools_count", tools.length);
					}
					span.setStatus({ code: SpanStatusCode.OK });
					return result;
				} catch (e) {
					span.recordException(e as Error);
					span.setStatus({ code: SpanStatusCode.ERROR, message: String(e) });
					throw e;
				} finally {
					span.end();
				}
			},
		);
	};
}

File: packages/sdk-python/tracelane/instrumentations/mcp.py
"""MCP (Model Context Protocol) instrumentation for Tracelane.

Wraps MCP client's call_tool and read_resource methods to emit OTel spans.
MCP is the most-attacked agent surface (rug-pull detection, PR1 guardrail)
and the most under-built observability layer. This adapter provides
the foundation for Tracelane's MCP hash watcher (PP-PR1).

Example::

    from mcp import Client
    from tracelane.instrumentations.mcp import instrument_mcp

    client = Client()
    instrument_mcp(client)
    result = await client.call_tool("my_tool", {"arg": "value"})
    # Span emitted with mcp.tool_name and argument count
"""

from __future__ import annotations

from typing import Any

from opentelemetry import trace
from opentelemetry.trace import SpanKind, StatusCode

_tracer = trace.get_tracer("tracelane.mcp", "0.1.0")


def instrument_mcp(client: Any) -> None:
    """Wrap an MCP Client instance to emit OTel spans.

    Instruments call_tool() and read_resource() if present.

    Args:
        client: An mcp.Client or mcp.ClientSession instance.

    Note:
        Tool arguments are counted but never captured as span attributes
        to avoid inadvertent PII capture. The tool name is captured.
    """
    if hasattr(client, "call_tool"):
        _patch_call_tool(client)
    if hasattr(client, "read_resource"):
        _patch_read_resource(client)


def _patch_call_tool(client: Any) -> None:
    original = client.call_tool

    async def _patched(tool_name: Any, arguments: Any = None, **kwargs: Any) -> Any:
        arg_count = len(arguments) if isinstance(arguments, dict) else 0
        with _tracer.start_as_current_span(
            "mcp.call_tool",
            kind=SpanKind.CLIENT,
            attributes={
                "gen_ai.provider.name": "mcp",
                "mcp.tool_name": str(tool_name),
                "mcp.argument_count": arg_count,
            },
        ) as span:
            try:
                result = await original(tool_name, arguments, **kwargs)
                is_error = getattr(result, "isError", False)
                span.set_attribute("mcp.is_error", bool(is_error))
                content = getattr(result, "content", None)
                if isinstance(content, list):
                    span.set_attribute("mcp.content_count", len(content))
                span.set_status(StatusCode.ERROR if is_error else StatusCode.OK)
                return result
            except Exception as exc:
                span.record_exception(exc)
                span.set_status(StatusCode.ERROR, str(exc))
                raise

    client.call_tool = _patched


def _patch_read_resource(client: Any) -> None:
    original = client.read_resource

    async def _patched(uri: Any, **kwargs: Any) -> Any:
        with _tracer.start_as_current_span(
            "mcp.read_resource",
            kind=SpanKind.CLIENT,
            attributes={
                "gen_ai.provider.name": "mcp",
                "mcp.resource_uri": str(uri),
            },
        ) as span:
            try:
                result = await original(uri, **kwargs)
                contents = getattr(result, "contents", None)
                if isinstance(contents, list):
                    span.set_attribute("mcp.content_count", len(contents))
                span.set_status(StatusCode.OK)
                return result
            except Exception as exc:
                span.record_exception(exc)
                span.set_status(StatusCode.ERROR, str(exc))
                raise

    client.read_resource = _patched
